src/Nethermind/Nethermind.Core/RateLimiter.cs

```diff
@@ -0,0 +1,66 @@
+// SPDX-FileCopyrightText: 2023 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Nethermind.Core;
+
+/// <summary>
+/// Simple rate limiter that limits rate of event, by delaying the caller so that a minimum amount of time elapsed
+/// between event.
+/// </summary>
+public class RateLimiter
+{
+    private TimeSpan _delay;
+    private DateTimeOffset _nextSlot;
+    private SpinLock _spinLock = new();
+
+    public RateLimiter(int eventPerSec) : this(1.0 / eventPerSec)
+    {
+    }
+
+    private RateLimiter(double intervalSec)
+    {
+        _delay = TimeSpan.FromSeconds(intervalSec);
+        _nextSlot = DateTimeOffset.Now;
+    }
+
+    public bool IsThrottled()
+    {
+        return DateTimeOffset.Now < _nextSlot;
+    }
+
+    public async Task WaitAsync(CancellationToken ctx)
+    {
+        while (true)
```

Comment:
looks like use case for periodic timer

Reply:
Specifically I think he's referring to [WaitForNextTickAsync(CancellationToken)](https://learn.microsoft.com/en-us/dotnet/api/system.threading.periodictimer.waitfornexttickasync?view=net-7.0#system-threading-periodictimer-waitfornexttickasync(system-threading-cancellationtoken))

Reply:
Seems to be single consumer only.